### ISSUE TITLE
Make general mechanism for hiding mouse cursor when running exhibits  

### ISSUE THREAD

**Issue by malex984** [IMAGINARY/IMAGINARY.OS#3]
IDEA: Make it similar to _Close-Button_ setting

**Comment by malex984** [comment]
Maybe same for the preferred/default language?

**Comment by elondaits** [comment]
Our electron kiosk browser should have this functionality as something configurable, because in some cases it could be run outside IMAGINARY OS.

... But perhaps it should  be an OS level setting as well, for some non-web apps? Or does the touch screen driver do that by itself?

As for the language configuration:

- Not everything is translatable yet
- Things that are translatable use different methods / policies / etc.
- It's a very complex domain and we're still just touching the surface (e.g. I'll eventually add support for localized versions of the same language, since argentine/uruguayan spanish is very different from the one from spain, chinese is different in the continent and Taiwan, etc... and then we have right-to-left languages, etc. Plus full localization support includes number formats, dates, etc.).

... So it requires a lot of work and thought. I wouldn't rush into doing anything complex.

**Comment by malex984** [comment]
I am planning to hide mouse within a wrapper `launcher.sh` script using some generic method (e.g. one of https://stackoverflow.com/questions/660613/how-do-you-hide-the-mouse-pointer-under-linux-x11).

**Comment by malex984** [comment]
I think that most exhibits will require new individual starter scripts to control their language facilities depending on the current user's settings. And we probably would not want to extend exhibits with language support (by ourselves)...
I agree that it is a complex thing.

**Comment by elondaits** [comment]
We should make a spreadsheet with all the programs in IMAGINARY OS and their status regarding i18n and l10n. Like if it requires translation, if it is translated, etc. so we know what we have on our hands. I can make it if you give me (or point me to) a list of the software included in IMAGINARY OS.

... Then we can plan. For web apps I can probably add translation support to the ones that don't have by using my i18n4js library, which I'll also integrate to MM when I have time. i18n4js gets the language from the URL and loads the translation from external JSON files.

**Comment by andreasdanielmatt** [comment]
yes, this is importnat (just to have an overview, apart from Rumanian we might get Italian now soon -
 for the exhibition planned in Genova - and if the Brasil projects for next year work out we will do a bigger Portuguese/Brasilian update - important big languages). Here is an "old" file with some kind of status (but definitely out-dated). I think Daniel once compiled it. This old Wiki will be "turned off" soon, so we will put thinks into Confluence or keep them in a spreadsheet (having links maybe in Confluence, so the spreadsheet does not get lost)

http://imaginary-manual.bedouine.org/w/index.php/Translations

**Comment by porst17** [comment]
Things like mouse pointer and default language should be a global setting (environment variable?) that can be overridden by the configuration of each individual program. On a touch screen, the mouse pointer should never be visible. On any other screen, the mouse pointer should always be visible or should be visible for a few seconds and then disappear until someone moves it again (hiding the mouse pointer in idle mode just looks better, see [`unclutter`](https://wiki.ubuntuusers.de/unclutter/)).

For the translations, each program would need a fallback language (usually english) that is used if the requested language is not available. This is how it's done in most localization systems: The programs asks the system for a translation of a given string in a given language. If it is not available, it will just return the given string again.

The kiosk browser can forward the global language setting to the web page, e.g. by using some url parameter.

**Comment by malex984** [comment]
@porst17 Yes, those global default settings should be user local settings just like with close button.
Note: i currently use `~/.show_close_button` with `SHOW_CLOSE_BUTTON=1` (or `=0`) instead of generating `/opt/launch.sh`.

@porst17 Do you think it may be worth having setting overrides for each application (e.g. `~/.show_close_button.TestApp`.

I agree about fallback language - this is why i have suggested to have the currently preferred language and a fallback default option (i.e. English).

**Comment by malex984** [comment]
I am not sure whether Kiosk browser can or should do anything about language settings since each web.
 app is quite different: some require to open corresponding `index_de.html` others accept a dynamic setting e.g. `.../applauncher2/?lang=ru`.

Which means that we need two things:
1. wrapper script (launcher.sh?) to read all the global settings (and maybe overrides) which can either handle settings itself (e.g. run `unclutter`) or would pass it into the application via corresponding application's interface (e.g. as with `dune-ash-ui` or via URL parameters for web.apps)
2. each web.app needs an entry point which can handle URL parameters according to that Application specifics...

**Comment by elondaits** [comment]
> For the translations, each program would need a fallback language (usually english) that is used if the requested language is not available. This is how it's done in most localization systems: The programs asks the system for a translation of a given string in a given language. If it is not available, it will just return the given string again.

@porst17 : The i18n library I used to make the CindyJS apps translatable (i18n4js) and that I'll eventually integrate with Sphere of the Earth 2 falls back to the default language (configurable) when the requested language does not exist or the requested string is not translated. I don't "return the given string" because I use string IDs (e.g. t("WELCOME_MSG") instead of t("Welcome to IMAGINARY")) because the latter is problematic: sometimes the same english phrase / word / fragment should be translated differently based on context (that's an issue with Drupal, which tries to do things this way). The library can be used for any web app, it's rather easy to integrate.

> The kiosk browser can forward the global language setting to the web page, e.g. by using some url parameter.

Browsers already send language settings: `navigator.language` gives the default language and `navigator.languages` the list of languages in order of preference. I have a TO DO to read this to set the default language.

Sending the language as a query string argument when loading the app as a way to transmit the OS's / browser's setting is not a good idea because if the app follows a link it'll lose the argument. Also if the app's URL includes a hashbang (#) you can't simply append a qs argument to the end. Modern dynamic apps done with JS frameworks like angular use hashbang for stateful URLs instead of QS.

... So I recommend using the existing navigator.language or defining our own JS API which we can inject in our electron browser.

But as I said before, it's a complex domain... it's not just a matter of passing language... there are "right to left" languages (hebrew, arabic), you also have to mind the number and date formats (US english uses m/d/y, british uses d/m/y), some apps can switch language dynamically while others need to be reloaded, and some support an internal language menu which you'd want to configure with reasonable choices or even hide. So we have to accommodate for all these complexity as we come across it and are able to, but we won't be able to figure out a 100% general solution right now.

**Comment by malex984** [comment]
@elondaits How exactly browsers get navigator.language? Is it something missing in our Kiosk browser ATM?

**Comment by porst17** [comment]
I think it is important to focus on the mechanism to pass default/global settings to apps. We should probably try to put as much of this state into the environment (like the language) and the apps just pick up what's there (e.g. this may be what finally lands in `navigator.language`). I don't think we should invent environment variables that are specific to IMAGINARY OS because we would need to add IMAGINARY OS specific code to all apps. But I am OK with small starter scripts that modify app configuration files (e.g. for hiding mouse pointer).

**Comment by elondaits** [comment]
@malex984 : Chrome/Firefox get it through their configuration screen (chrome://settings/) I don't know if Electron has the same and/or offers the ability to set that configuration through a file, argument, etc. 

@porst17 : I don't like the idea of a script modifying the config file of a web app (e.g. the new app launcher) because then it takes control from the admin who would ideally edit that file (I made it as friendly to edit as possible) and adds a huge potential for errors. 

I'd rather get the defaults into electron somehow (ideally existing variables, like navigator.language, which work even when not using electron) where apps we do ourselves can look for them (and if they exist, use them). It's no use to do an ugly kludge for a codebase we control 100% instead of a clean interface. If the app updates the ugly kludge breaks... fixing it involves the app developer and the script developer, instead of being testable code that the developer managers. In the case of legacy apps or things we didn't do ourselves and have limited control over it's OK to do it in any easy way.

**Comment by malex984** [comment]
Ok, I will check how to incorporate/support `navigator.language` in Kiosk Browser.
Linking this to the super issue https://github.com/hilbert/hilbert-docker-images/issues/51
